ArcGISRuntimeSDKQt_CppSamples/EditData/SnapGeometryEdits/SnapGeometryEdits.cpp

```diff
@@ -0,0 +1,335 @@
+  case GeometryEditorMode::PointMode:
+    m_geometryEditor->start(Esri::ArcGISRuntime::GeometryType::Point);
+    break;
+
+  case GeometryEditorMode::MultipointMode:
+    m_geometryEditor->start(Esri::ArcGISRuntime::GeometryType::Multipoint);
+    break;
+
+  case GeometryEditorMode::PolylineMode:
+    m_geometryEditor->start(Esri::ArcGISRuntime::GeometryType::Polyline);
+    break;
+
+  case GeometryEditorMode::PolygonMode:
+    m_geometryEditor->start(Esri::ArcGISRuntime::GeometryType::Polygon);
+    break;
+
+  default:
+    break;
+  }
+  emit geometryEditorStartedChanged();
+}
+
+// Stops the GeometryEditor and append the new graphic to the graphics overlay
+void SnapGeometryEdits::stopEditor()
+{
+  const Geometry geometry = m_geometryEditor->stop();
+  emit geometryEditorStartedChanged();
+
+  if (m_editingGraphic)
+  {
+    m_editingGraphic->setGeometry(geometry);
+
+    m_editingGraphic->setVisible(true);
+    m_editingGraphic = nullptr;
+
+    return;
+  }
+
+  Symbol* geometrySymbol = nullptr;
+  switch (geometry.geometryType())
+  {
+  case GeometryType::Point:
+    geometrySymbol = m_pointSymbol;
+    break;
+
+  case GeometryType::Multipoint:
+    geometrySymbol = m_multiPointSymbol;
+    break;
+
+  case GeometryType::Polyline:
+    geometrySymbol = m_lineSymbol;
+    break;
+
+  case GeometryType::Polygon:
+    geometrySymbol = m_polygonSymbol;
+    break;
+
+  default:
+    return;
+  }
+
+  m_graphicsOverlay->graphics()->append(new Graphic(geometry, geometrySymbol, new QObject(this)));
+}
+
+// Deletes the selected element
+void SnapGeometryEdits::deleteSelection()
+{
+  m_geometryEditor->deleteSelectedElement();
+  emit canUndoChanged();
+}
+
+// Reverts the last event on the geometry editor
+void SnapGeometryEdits::editorUndo()
+{
+  m_geometryEditor->undo();
+  emit canUndoChanged();
+}
+
+void SnapGeometryEdits::displaySnapSources()
+{
+  // create lists for displaying the snap sources in snap settings
+  QList<Esri::ArcGISRuntime::SnapSourceSettings*> snapSourceList;
+  if (m_geometryEditor->snapSettings()->sourceSettings().empty())
+  {
+    m_geometryEditor->snapSettings()->syncSourceSettings();
+    snapSourceList = m_geometryEditor->snapSettings()->sourceSettings();
+
+    for (int index = 0; index < snapSourceList.size()-1; ++index)
+    {
+      FeatureLayer *featureLayer = static_cast<FeatureLayer*>(snapSourceList[index]->source());
+      if (featureLayer->featureTable()->geometryType() == GeometryType::Point)
+      {
+        m_pointLayers.append(featureLayer->name());
+        m_pointSourceCheckedState.append(snapSourceList[index]->isEnabled());
+        m_pointSourceList.append(snapSourceList[index]);
+      }
+      else if (featureLayer->featureTable()->geometryType() == GeometryType::Polyline)
+      {
+        m_polylineLayers.append(featureLayer->name());
+        m_polylineSourceCheckedState.append(snapSourceList[index]->isEnabled());
+        m_polylineSourceList.append(snapSourceList[index]);
+      }
```

Comment:
I recommend rewriting this to make sure things are valid.
```suggestion
      if (FeatureLayer* featureLayer = dynamic_cast<FeatureLayer*>(snapSourceList[index]->source()); featureLayer != nullptr)
      {
        const GeometryType featureTableGeomType = featureLayer->featureTable()->geometryType();
        if (featureTableGeomType == GeometryType::Point)
        {
          m_pointLayers.append(featureLayer->name());
          m_pointSourceCheckedState.append(snapSourceList[index]->isEnabled());
          m_pointSourceList.append(snapSourceList[index]);
        }
        else if (featureTableGeomType == GeometryType::Polyline)
        {
          m_polylineLayers.append(featureLayer->name());
          m_polylineSourceCheckedState.append(snapSourceList[index]->isEnabled());
          m_polylineSourceList.append(snapSourceList[index]);
        }
      }
```

Make sure all the syntax actually works.